polus-ftl-label-plugin/src/ftl.pyx

```diff
@@ -0,0 +1,966 @@
+
+    int _mm_movemask_epi8(__m128i a) nogil
+
+    long long _mm_cvtsi128_si64(__m128i a) nogil
+
+    __m128i _mm_setzero_si128() nogil
+
+    __m128i _mm_slli_si128(__m128i a,int b) nogil
+
+    __m128i _mm_loadu_si128(__m128i* d) nogil
+
+    __m128i _mm_cmpgt_epi8(__m128i a,__m128i b) nogil
+
+    __m128i _mm_xor_si128(__m128i a,__m128i b) nogil
+
+    __m128i _mm_srli_epi64(__m128i a,int i) nogil
+
+    __m128i _mm_shuffle_epi8(__m128i a,__m128i b) nogil
+
+    __m256i _mm256_add_epi16(__m256i a, __m256i b) nogil
+    
+    __m128i _mm_set_epi8(char e15,char e14,char e13,char e12,
+                         char e11,char e10,char e9, char e8,
+                         char e7, char e6, char e5, char e4,
+                         char e3, char e2, char e1, char e0) nogil
+
+    __m256i _mm256_set_epi8(char e31,char e30,char e29,char e28,
+                            char e27,char e26,char e25,char e24,
+                            char e23,char e22,char e21,char e20,
+                            char e19,char e18,char e17,char e16,
+                            char e15,char e14,char e13,char e12,
+                            char e11,char e10,char e9, char e8,
+                            char e7, char e6, char e5, char e4,
+                            char e3, char e2, char e1, char e0) nogil
+    
+    __m256i _mm256_set_epi64x(long long e3,long long e2,
+                              long long e1,long long e0) nogil
+
+    __m256i _mm256_set1_epi8(char a) nogil
+
+    __m256i _mm256_set1_epi16(short a) nogil
+
+    __m256i _mm256_set1_epi32(int a) nogil
+
+    __m256i _mm256_set1_epi64x(long long a) nogil
+
+    __m256i _mm256_srlv_epi64(__m256i a, __m256i b) nogil
+
+    __m256i _mm256_and_si256(__m256i a, __m256i b) nogil
+
+    __m256i _mm256_shuffle_epi8(__m256i a, __m256i b) nogil
+
+    __m256i _mm256_add_epi16(__m256i a, __m256i b) nogil
+
+    __m256i _mm256_add_epi32(__m256i a, __m256i b) nogil
+
+    void _mm256_storeu_si256(__m256i* a,__m256i b) nogil
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+@cython.nonecheck(False)
+cdef np.ndarray run_length_encode_16(unsigned char [:] image,tuple shape):
+    """run_length_encode Create run length encoded pixels
+
+    This function will run on a matrix with arbitrary number of dimensions.
+    It is best to use the appropriate run length encoding algorithm for a
+    specified number of dimensions if it exists.
+
+    This function finds consecutive pixels along the final dimension
+    that have non-zero values. Then, the first and last dimension index
+    along with the other coordinates are stored in a numpy array. This
+    results in a 2d numpy array with ndim+1 columns for each column object:
+    [...,last_dim_index_start,last_dim_index_end]
+
+    The input image must be a linear array (i.e. use reshape(-1)).
+    The input image must be np.uint8 or bool data type.
+
+    Note: This function is the main bottleneck in the labeling algorithm.
```

Comment:
There is a multi-threaded version of Cython that can be implemented with addition of just a few lines of code. If the performance of code is still slow, you might be interested in it. It is explained here:https://software.intel.com/content/www/us/en/develop/articles/thread-parallelism-in-cython.html

Reply:
Thanks for the reference. I'll have to check that out and wasn't aware of it. Cython also has native integration with OpenMP. When I was writing my original C code, I tried to make it multithreaded but it didn't result in very significant increases in speed. My assumption was that I may have been more or less bus speed limited because the calculations once the data is sent to cache are fairly fast, so most of the time was being spent sending data to or from cache. However, I also may not have known what I was doing.